helix-core/src/movement.rs

```diff
@@ -1,184 +1,254 @@
-
-        if categorize(ch) != categorize(prev) {
-            begin -= 1;
-        }
-
-        // return if not skip while?
-        skip_over_prev(slice, &mut begin, |ch| ch == '\n');
+pub fn move_prev_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        (range.head > 0 && range.head.inside(slice)).then(|| {
+            let new_anchor = if range.head.saturating_sub(1).is_boundary(slice) {
+                (range.head.saturating_sub(1)).backwards_skip_newlines(slice)
+            } else {
+                range.head.backwards_skip_newlines(slice)
+            };
+            let new_head = range.head.saturating_sub(1).start_of_block(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        end = begin;
+pub fn move_next_word_end(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
+            let new_anchor = if range.head.is_boundary(slice) {
+                after_head
+            } else {
+                range.head.skip_newlines(slice)
+            };
+            let new_head = (range.head + 1).end_of_word(slice);
+            Some(Range::new(new_anchor, new_head))
+        })?
+    };
+    (0..count).fold(range, |range, _| movement(range).unwrap_or(range))
+}
 
-        with_end = skip_over_prev(slice, &mut end, char::is_whitespace);
+// ---- util ------------
 
-        // refetch
-        let ch = slice.char(end);
+/// Private helpers to help manipulate slice indices
+pub trait SliceIndexHelpers {
+    fn outside(&self, slice: RopeSlice) -> bool;
+    fn inside(&self, slice: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_boundary(&self, slice: RopeSlice) -> bool;
+    fn category(&self, slice: RopeSlice) -> Option<Category>;
+    /// Returns the start of a word/punctuation group followed by any amount of whitespace.
+    fn start_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group followed by any amount of whitespace.
+    fn end_of_block(&self, slice: RopeSlice) -> Self;
+    /// Returns the end of a word/punctuation group.
+    fn end_of_word(&self, slice: RopeSlice) -> Self;
+    fn skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn backwards_skip_newlines(&self, slice: RopeSlice) -> Self;
+    fn is_whitespace(&self, slice: RopeSlice) -> bool;
+    fn is_end_of_line(&self, slice: RopeSlice) -> bool;
+}
 
-        if is_word(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_punctuation);
-        }
+impl SliceIndexHelpers for usize {
```

Comment:
I personally think this `impl` makes the code harder to understand since it's `usize` and we shouldn't `impl` it for `usize`, maybe a specific type yes.

Reply:
Any particular reason why you wouldn't want to impl for `usize`? (Particularly for private traits like these).

Reply:
I find these parts made the upper parts harder to understand.

Reply:
I think flipping it will make it easier to understand. I mean rather `impl for usize`, `impl for Rope` will result in slightly longer code but the flow will be more understandable. Rather than `wd`, we have `dw`.